main.cpp

```diff
@@ -19,13 +19,31 @@ using namespace clang::tooling;
 
 class CastCallBack : public MatchFinder::MatchCallback {
 public:
-    CastCallBack(Rewriter& rewriter) {
-        // Your code goes here
-    };
+    CastCallBack(Rewriter& rewriter) : r(rewriter) {};
 
     void run(const MatchFinder::MatchResult &Result) override {
-        // Your code goes here
+        const StringRef castText("static_cast");
+
+        if (const auto* expr = Result.Nodes.getNodeAs<CStyleCastExpr>("cast")) {
+            const auto& SM = *Result.SourceManager;
+
+            const auto replaceRange = CharSourceRange::getCharRange(expr->getLParenLoc(), expr->getSubExprAsWritten()->getBeginLoc());
+            const auto dstTypeString = Lexer::getSourceText(CharSourceRange::getTokenRange(
+                    expr->getLParenLoc().getLocWithOffset(1), expr->getRParenLoc().getLocWithOffset(-1)), SM, LangOptions());
+
+            std::string newCastText = (castText + "<" + dstTypeString + ">").str();
+
+            const auto* subExpr = expr->getSubExprAsWritten()->IgnoreImpCasts();
```

Comment:
Please describe what happens if we change the code as follows:  
```suggestion
            const auto* subExpr = expr->getSubExprAsWritten();
```
What is the purpose of using the `IgnoreImpCasts()` method?

Reply:
I saw it in [complicated implementation](https://github.com/llvm-mirror/clang-tools-extra/blob/5c40544fa40bfb85ec888b6a03421b3905e4a4e7/clang-tidy/google/AvoidCStyleCastsCheck.cpp#L137). I'm not sure what is it but I left it like that just in case.

I thought that it is for cases `int x = 0; int b = (int)b;` but it was casted anyway

Reply:
If you try to print the AST of the `test.cpp` example, you will see that you have several `ImplicitCastExpr` nodes:
![image](https://user-images.githubusercontent.com/76072618/119860568-20cb0700-bf1f-11eb-8e3d-2a533baa265f.png)

Depending on which approaches you use to get the subexpression node, you will get different result nodes for the `test.cpp` example:

Code:
```cpp
// getSubExpr() retrieves the first cast subexpression
llvm::outs() << "getSubExpr(): " << castNode->getSubExpr()->getStmtClassName();

// getSubExprAsWritten() retrieves the cast subexpression as it was written in the source
// code, looking through any implicit casts or other intermediate nodes
// introduced by semantic analysis.
llvm::outs() << "getSubExprAsWritten(): " << castNode->getSubExprAsWritten()->getStmtClassName();

// getSubExpr() retrieves the first cast subexpression.
// ignoreImpCasts() skips past any implicit casts which might surround this expression until
// reaching a fixed point. It skips the following AST nodes: ImplicitCastExpr, FullExpr.
llvm::outs() << "getSubExpr()->ignoreImpCasts(): " << castNode->getSubExpr()->IgnoreImpCasts()->getStmtClassName();
```
Output:
```bash
 getSubExpr(): ImplicitCastExpr 
 getSubExprAsWritten(): DeclRefExpr  
 getSubExpr()->ignoreImpCasts(): DeclRefExpr
```

Reply:
I understood, thanks!